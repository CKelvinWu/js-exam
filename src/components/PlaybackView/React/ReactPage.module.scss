.app {
  display: flex;
  flex-direction: row;
}

.rightWidget {
  height: 33.3%;
  width: 100%;
  border: 1px solid #eee;
}

.label {
	position: absolute;
  margin-top: 5px;
  right: 17px;
	z-index: 10;
	padding: 2px 12px;
	color: white;
	background-color: rgba($color: #000000, $alpha: 0.5);
}