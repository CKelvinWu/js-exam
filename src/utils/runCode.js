import _ from 'underscore';
import sinon from 'sinon';
import vm from 'vm';
import spy from './spy';
import getPatchedTape from './tape';
import { calendarFormat } from '../../node_modules/moment';

const wrapCode = (code = '') => {
  return code.replace(/for *\(.*\{|while *\(.*\{|do *\{/g, (loopHead) => {
    return `let count = 0;
      const detector = (c) => {
        if (c > 100000) {
          throw new Error('infinite');
        }
      }
      ${loopHead}
      detector(count++);
    `;
  });
}

const runCode = (code, wrappedConsole) => {
  const test = getPatchedTape();
  // should hijack setTimeout before pass to sandbox
  const clock = sinon.useFakeTimers();
  const sandbox = {
    setTimeout: window.setTimeout, // need to be passed also...
    console: wrappedConsole,
    sinon,   
    describe: test,
    test,
    clock,
    spy
  };
  const script = new vm.Script(wrapCode(code));
  const context = new vm.createContext(sandbox);
  script.runInContext(context);
  clock.restore();
}
const debouncedRunCode = _.debounce(runCode, 200);

export default debouncedRunCode ;