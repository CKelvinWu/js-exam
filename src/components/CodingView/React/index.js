import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { withRouter } from 'react-router-dom';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import Grid from 'components/Grid';
import GridItem from 'components/Grid/GridItem';
import ConsoleWidget from 'components/Widgets/ConsoleWidget';
import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  static propTypes = {
    isExaming: PropTypes.bool,
    code: PropTypes.string,
    compiledCode: PropTypes.string,
    consoleMsg: PropTypes.array,
    handleCodeChange: PropTypes.func,
    wrappedConsole: PropTypes.object,
    resetConsole: PropTypes.func,
  };

  controlHeight = 70;

  componentDidMount() {
    const { compiledCode, wrappedConsole, resetConsole } = this.props;
    resetConsole();
    debouncedRunCode({ code: compiledCode, wrappedConsole });
  }

  shouldComponentUpdate(nextProps) {
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode, wrappedConsole, resetConsole } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      resetConsole();
      // WARNING: This is not debounced
      debouncedRunCode({ code: compiledCode, wrappedConsole });
    }
    return true;
  }

  render() {
    const { isExaming, code, handleCodeChange, consoleMsg } = this.props;
    const layout = [
      {
        key: 'code',
        x: 0,
        y: 0,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 + 100,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'answer',
        x: 0,
        y: 1,
        width: window.innerWidth / 2 + 50,
        height: window.innerHeight / 2 - 150,
        minWidth: 100,
        maxWidth: 700,
      },
      {
        key: 'result',
        x: 1,
        y: 0,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
      {
        key: 'console',
        x: 1,
        y: 1,
        width: window.innerWidth / 2 - 50,
        height: (window.innerHeight - this.controlHeight) / 2 - 50,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
    ];
    return (
      <div className={styles.app}>
        <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
          <GridItem key="code" label={GRID_LABEL_REACT.code}>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              readOnly={!isExaming}
              mode="jsx"
              theme="monokai"
            />
          </GridItem>
          <GridItem key="answer" label={GRID_LABEL_REACT.answer}>
            {isExaming ? <AnswerWidget /> : null}
          </GridItem>
          <GridItem key="result" label={GRID_LABEL_REACT.result}>
            {isExaming ? <ResultWidget /> : null}
          </GridItem>
          <GridItem key="console" label={GRID_LABEL_REACT.console}>
            {isExaming ? <ConsoleWidget data={consoleMsg} /> : null}
          </GridItem>
        </Grid>
      </div>
    );
  }
}

export default withRouter(ReactPage);
