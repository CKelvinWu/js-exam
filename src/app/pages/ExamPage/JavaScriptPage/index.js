import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import 'brace';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import Grid from 'app/components/Grid';
import GridItem from 'app/components/Grid/GridItem';
import ConsoleWidget from 'app/components/Widgets/ConsoleWidget';
import CodeWidget from 'app/components/Widgets/CodeWidget';
import TestWidget from 'app/components/Widgets/TestWidget';
import TapeWidget from 'app/components/Widgets/TapeWidget';
import ControlWidget from '../ControlWidget';

import debouncedRunCode from 'app/utils/runCode';

import styles from './JavaScriptPage.module.scss';

class JavaScriptPage extends Component {
  constructor(props) {
    super(props);
    this.controlHeight = 70;
  }

  componentDidMount() {
    const { compiledCode, wrappedConsole, resetConsole, addTape } = this.props;
    resetConsole();
    debouncedRunCode({ code: compiledCode, wrappedConsole, onTapeUpdate: addTape });
  }

  shouldComponentUpdate(nextProps) {
    const { compiledCode: previousCompiledCode, resetConsole, addTape, resetTape } = this.props;
    const { compiledCode, wrappedConsole } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      resetConsole();
      resetTape();
      debouncedRunCode({ code: compiledCode, wrappedConsole, onTapeUpdate: addTape });
    }
    return true;
  }

  render() {
    const {
      code,
      test,
      handleCodeChange,
      tape,
      console: _console,
      onReset,
    } = this.props;
    const layout = [
      {
        key: 'code', x: 0, y: 0, width: window.innerWidth / 2, height: window.innerHeight / 2, minWidth: 100, minHeight: 100, maxWidth: 700, maxHeight: 500
      },
      {
        key: 'test', x: 0, y: 1, width: window.innerWidth / 2, height: window.innerHeight / 2, minWidth: 100, maxWidth: 700
      },
      {
        key: 'control', x: 1, y: 0, width: window.innerWidth / 2, height: this.controlHeight, static: true
      },
      {
        key: 'tape', x: 1, y: 1, width: window.innerWidth / 2, height: (window.innerHeight - this.controlHeight) / 2, minWidth: 100, minHeight: 100, maxWidth: 700, maxHeight: 500
      },
      {
        key: 'console', x: 1, y: 2, width: window.innerWidth / 2, height: (window.innerHeight - this.controlHeight) / 2, minWidth: 100, minHeight: 100, maxWidth: 700, maxHeight: 500
      },
    ];
    return (
      <div className={styles.app}>
        <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
          <GridItem key="code">
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              mode="javascript"
              theme="monokai"
            />
          </GridItem>
          <GridItem key="test">
            <TestWidget data={test} />
          </GridItem>
          <GridItem key="control">
            <ControlWidget
              onReset={() => onReset('javascript')}
            />
          </GridItem>
          <GridItem key="tape">
            <TapeWidget data={tape} />
          </GridItem>
          <GridItem key="console">
            <ConsoleWidget data={_console} />
          </GridItem>
        </Grid>
      </div>
    );
  }
}

export default withRouter(JavaScriptPage);
