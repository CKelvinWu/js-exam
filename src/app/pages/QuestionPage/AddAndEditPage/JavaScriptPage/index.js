import React, { Component } from 'react';
import 'brace';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import { Spin } from 'antd';

import Grid from 'app/components/Grid';
import GridItem from 'app/components/Grid/GridItem';
import CodeWidget from 'app/components/Widgets/CodeWidget';
import TestWidget from 'app/components/Widgets/TestWidget';
import TapeWidget from 'app/components/Widgets/TapeWidget';

import debouncedRunCode from 'app/utils/runCode';

import TagWidget from '../../TagWidget';
import styles from './JavaScriptPage.module.scss';

class JavaScriptPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      tape: []
    };
  }

  componentDidMount() {
    const { compiledCode } = this.props;
    debouncedRunCode({ code: compiledCode, onTapeUpdate: this.addTape });
  }

  shouldComponentUpdate(nextProps) {
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      this.setState({ tape: [] }, () => {
        debouncedRunCode({ code: compiledCode, onTapeUpdate: this.addTape });
      });
    }
    return true;
  }

  addTape = data => {
    const { tape } = this.state;
    this.setState({
      tape: [...tape, data]
    });
  };

  render() {
    const { tape } = this.state;
    const {
      isLoading,
      test,
      code,
      tags,
      onTagUpdate,
      handleCodeChange,
      handleTestChange } = this.props;
    const layout = [
      {
        key: 'code',
        x: 0,
        y: 0,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500
      },
      {
        key: 'test',
        x: 0,
        y: 1,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        maxWidth: 700
      },
      {
        key: 'tape',
        x: 1,
        y: 0,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500
      },
      {
        key: 'tag',
        x: 1,
        y: 1,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500
      }
    ];
    return (
      <div className={styles.app}>
        <Spin spinning={isLoading} size="large">
          <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
            <GridItem key="code">
              <CodeWidget
                handleCodeChange={handleCodeChange}
                data={code}
                mode="javascript"
                theme="monokai"
              />
            </GridItem>
            <GridItem key="test">
              <TestWidget
                handleCodeChange={handleTestChange}
                data={test}
                readOnly={false}
              />
            </GridItem>
            <GridItem key="tape">
              <TapeWidget data={tape} />
            </GridItem>
            <GridItem key="tag">
              <TagWidget data={tags} onTagUpdate={onTagUpdate} />
            </GridItem>
          </Grid>
        </Spin>
      </div>
    );
  }
}

export default JavaScriptPage;
