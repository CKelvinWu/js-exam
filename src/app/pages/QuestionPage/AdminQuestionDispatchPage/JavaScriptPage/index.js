import React, { Component } from 'react';
import 'brace';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import { Spin } from 'antd';

import Grid from 'app/components/Grid';
import GridItem from 'app/components/Grid/GridItem';
import CodeWidget from 'app/components/Widgets/CodeWidget';
import TestWidget from 'app/components/Widgets/TestWidget';
import TapeWidget from 'app/components/Widgets/TapeWidget';

import debouncedRunCode from 'app/utils/runCode';

import ControlWidget from '../ControlWidget';
import TagWidget from '../../TagWidget';
import styles from './JavaScriptPage.module.scss';

class JavaScriptPage extends Component {

  controlHeight = 70;

  componentDidMount() {
    const { compiledCode, addTape } = this.props;
    debouncedRunCode({ code: compiledCode, onTapeUpdate: addTape });
  }

  render() {
    const {
      onChangeCategory,
      onDispatchQuestion,
      onChangeQuestion,
      onTagUpdate,
      categoryIndex,
      questionIndex,
      handleCodeChange,
      code,
      test,
      tape,
      tags,
      questionList,
      isLoading
    } = this.props;
    const layout = [
      {
        key: 'code',
        x: 0,
        y: 0,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500
      },
      {
        key: 'test',
        x: 0,
        y: 1,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        maxWidth: 700
      },
      {
        key: 'control',
        x: 1,
        y: 0,
        width: window.innerWidth / 2,
        height: this.controlHeight,
        static: true
      },
      {
        key: 'tape',
        x: 1,
        y: 1,
        width: window.innerWidth / 2,
        height: (window.innerHeight - this.controlHeight) / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500
      },
      {
        key: 'tag',
        x: 1,
        y: 2,
        width: window.innerWidth / 2,
        height: (window.innerHeight - this.controlHeight) / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500
      }
    ];
    return (
      <div className={styles.app}>
        <Spin spinning={isLoading} size="large">
          <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
            <GridItem key="code">
              <CodeWidget
                handleCodeChange={handleCodeChange}
                data={code}
                mode="javascript"
                theme="monokai"
              />
            </GridItem>
            <GridItem key="test">
              <TestWidget
                data={test}
                readOnly={false}
              />
            </GridItem>
            <GridItem key="control">
              <ControlWidget
                type="javascript"
                onDispatchQuestion={onDispatchQuestion}
                onChangeCategory={onChangeCategory}
                onChangeQuestion={onChangeQuestion}
                categoryIndex={categoryIndex}
                questionIndex={questionIndex}
                questionList={questionList}
              />
            </GridItem>
            <GridItem key="tape">
              <TapeWidget data={tape} />
            </GridItem>
            <GridItem key="tag">
              <TagWidget data={tags} onTagUpdate={onTagUpdate} />
            </GridItem>
          </Grid>
        </Spin>
      </div>
    );
  }
}

export default JavaScriptPage;
