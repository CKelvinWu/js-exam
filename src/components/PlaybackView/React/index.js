import React, { Component } from 'react';
import PropTypes from 'prop-types';
import 'brace';
import 'brace/mode/jsx';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import { Spin } from 'antd';

import CodeWidget from 'components/Widgets/CodeWidget';
import ResultWidget from 'components/Widgets/ResultWidget';
import AnswerWidget from 'components/Widgets/AnswerWidget';

import debouncedRunCode from 'utils/runCode';
import { REACT as GRID_LABEL_REACT } from 'utils/gridLabel';

import styles from './ReactPage.module.scss';

class ReactPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      height: props.height,
      widthCode: props.width,
      widthTests: props.width,
    };
  }

  async componentDidMount() {
    const { compiledCode, addTape } = this.props;

    this.setState({
      height: `${window.innerHeight - 160}px`,
      widthCode: `${window.innerWidth * 0.45}px`,
      widthTests: `${window.innerWidth * 0.3}px`,
    });
    debouncedRunCode({ code: compiledCode, onTapeUpdate: addTape });
  }

  shouldComponentUpdate(nextProps) {
    
    const { compiledCode: previousCompiledCode } = this.props;
    const { compiledCode } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      debouncedRunCode({ code: compiledCode });
    }
    return true;
  }

  render() {
    this.codeWidgetstyle = {
      height: this.state.height,
      width: this.state.widthCode,
      position: 'relative',
    };
    this.testWidgetStyle = {
      height: this.state.height,
      width: this.state.widthTests,
      position: 'relative',
    };
    const { handleCodeChange, test, code } = this.props;
    return (
      <React.Fragment>
        <div className={styles.app}>
          <div style={this.codeWidgetstyle}>
            <div className={styles.label}>Code</div>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              mode="jsx"
              theme="monokai"
              readOnly
            />
          </div>
          <div style={this.testWidgetStyle}>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.test}</div>
              <CodeWidget data={test} mode="jsx" theme="textmate" readOnly />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.answer}</div>
              <AnswerWidget />
            </div>
            <div className={styles.rightWidget}>
              <div className={styles.label}>{GRID_LABEL_REACT.result}</div>
              <ResultWidget />
            </div>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

ReactPage.propTypes = {
  compiledCode: PropTypes.string,
  code: PropTypes.string,
  height: PropTypes.string,
  width: PropTypes.string,
  test: PropTypes.string,
  handleCodeChange: PropTypes.func,
  addTape: PropTypes.func,
};

export default ReactPage;
