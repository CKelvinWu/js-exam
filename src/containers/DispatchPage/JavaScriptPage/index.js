import React, { Component } from 'react';
import 'brace';
import 'brace/mode/javascript';
import 'brace/theme/textmate';
import 'brace/theme/monokai';

import { Spin } from 'antd';

import Grid from 'components/Grid';
import GridItem from 'components/Grid/GridItem';
import CodeWidget from 'components/Widgets/CodeWidget';
import TestWidget from 'components/Widgets/TestWidget';
import TapeWidget from 'components/Widgets/TapeWidget';

import debouncedRunCode from 'utils/runCode';
import { JAVASCRIPT as GRID_LABEL_JAVASCRIPT  } from 'utils/gridLabel';

import styles from './JavaScriptPage.module.scss';

class JavaScriptPage extends Component {
  componentDidMount() {
    const { compiledCode, addTape } = this.props;
    debouncedRunCode({ code: compiledCode, onTapeUpdate: addTape });
  }

  shouldComponentUpdate(nextProps) {
    const {
      compiledCode: previousCompiledCode,
      addTape,
      resetTape,
    } = this.props;
    const { compiledCode } = nextProps;
    if (previousCompiledCode !== compiledCode) {
      resetTape();
      debouncedRunCode({ code: compiledCode, onTapeUpdate: addTape });
    }
    return true;
  }

  render() {
    const {
      onTagUpdate,
      handleCodeChange,
      code,
      test,
      tape,
    } = this.props;
    const layout = [
      {
        key: 'code',
        x: 0,
        y: 0,
        width: window.innerWidth / 2,
        height: window.innerHeight,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: window.innerHeight,
      },
      {
        key: 'test',
        x: 1,
        y: 0,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        maxWidth: 700,
      },
      {
        key: 'tape',
        x: 1,
        y: 1,
        width: window.innerWidth / 2,
        height: window.innerHeight / 2,
        minWidth: 100,
        minHeight: 100,
        maxWidth: 700,
        maxHeight: 500,
      },
    ];
    return (
      <div className={styles.app}>
        <Grid layout={layout} totalWidth="100%" totalHeight="100%" autoResize>
          <GridItem key="code" label={GRID_LABEL_JAVASCRIPT.code}>
            <CodeWidget
              handleCodeChange={handleCodeChange}
              data={code}
              mode="javascript"
              theme="monokai"
              readOnly
            />
          </GridItem>
          <GridItem key="test" label={GRID_LABEL_JAVASCRIPT.test}>
            <TestWidget data={test} readOnly />
          </GridItem>
          <GridItem key="tape" label={GRID_LABEL_JAVASCRIPT.tape}>
            <TapeWidget data={tape} />
          </GridItem>
        </Grid>
      </div>
    );
  }
}

export default JavaScriptPage;
