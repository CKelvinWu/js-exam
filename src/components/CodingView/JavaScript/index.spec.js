import React from 'react';
import { mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import sinon from 'sinon';
import * as runCodeUtils from '../../../utils/runCode';
import JavaScriptPage from './index';

describe('Coding View Javascript index', () => {
  it('should render Javascript coding view properly', () => {
    const spy = sinon.spy(runCodeUtils, 'default');
    const element = mount(
      <BrowserRouter>
        <JavaScriptPage
          isExaming
          code="const a = 3;"
          compiledCode="const a = 3;"
          test="abcde"
          consoleMsg={['1', '2']}
          tape={['1']}
          handleCodeChange={jest.fn()}
          wrappedConsole={{}}
          resetConsole={jest.fn()}
          addTape={jest.fn()}
          resetTape={jest.fn()}
        />
      </BrowserRouter>,
    );
    expect(element.render()).toMatchSnapshot();
    expect(spy.calledOnce).toBeTruthy();
  });
});
