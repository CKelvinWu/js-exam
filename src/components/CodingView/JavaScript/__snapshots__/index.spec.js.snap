// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Coding View Javascript index should render Javascript coding view properly 1`] = `
<div
  class="app"
>
  <div
    class="grid"
    style="width: 100%; height: 100%;"
  >
    <div
      class="border  horizontal"
      style="width: 567px; padding-right: 5px;"
    >
      <div
        class="border  vertical"
        style="width: 562px; height: 484px; padding-bottom: 5px;"
      >
        <div
          class="item"
        >
          <div
            class="label"
          >
            code
          </div>
          <div
            class=" ace_editor ace-monokai ace_dark code"
            id="brace-editor"
            style="width: 500px; height: 500px;"
          >
            <textarea
              autocapitalize="off"
              autocorrect="off"
              class="ace_text-input"
              spellcheck="false"
              style="opacity: 0;"
              wrap="off"
            />
            <div
              aria-hidden="true"
              class="ace_gutter"
            >
              <div
                class="ace_layer ace_gutter-layer ace_folding-enabled"
              />
              <div
                class="ace_gutter-active-line"
              />
            </div>
            <div
              class="ace_scroller"
            >
              <div
                class="ace_content"
              >
                <div
                  class="ace_layer ace_print-margin-layer"
                >
                  <div
                    class="ace_print-margin"
                    style="left: 4px; visibility: hidden;"
                  />
                </div>
                <div
                  class="ace_layer ace_marker-layer"
                />
                <div
                  class="ace_layer ace_text-layer"
                  style="padding: 0px 4px;"
                />
                <div
                  class="ace_layer ace_marker-layer"
                />
                <div
                  class="ace_layer ace_cursor-layer ace_hidden-cursors"
                >
                  <div
                    class="ace_cursor"
                  />
                </div>
              </div>
            </div>
            <div
              class="ace_scrollbar ace_scrollbar-v"
              style="display: none; width: 20px;"
            >
              <div
                class="ace_scrollbar-inner"
                style="width: 20px;"
              />
            </div>
            <div
              class="ace_scrollbar ace_scrollbar-h"
              style="display: none; height: 20px;"
            >
              <div
                class="ace_scrollbar-inner"
                style="height: 20px;"
              />
            </div>
            <div
              style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: hidden;"
            >
              <div
                style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
              />
              <div
                style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
              >
                XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="border "
        style="width: 562px; height: -484px;"
      >
        <div
          class="item"
        >
          <div
            class="label"
          >
            unit test
          </div>
          <div
            class="test-widget"
          >
            <div
              class=" ace_editor ace-tm"
              id="brace-editor"
              style="width: 500px; height: 500px;"
            >
              <textarea
                autocapitalize="off"
                autocorrect="off"
                class="ace_text-input"
                spellcheck="false"
                style="opacity: 0;"
                wrap="off"
              />
              <div
                aria-hidden="true"
                class="ace_gutter"
              >
                <div
                  class="ace_layer ace_gutter-layer ace_folding-enabled"
                />
                <div
                  class="ace_gutter-active-line"
                />
              </div>
              <div
                class="ace_scroller"
              >
                <div
                  class="ace_content"
                >
                  <div
                    class="ace_layer ace_print-margin-layer"
                  >
                    <div
                      class="ace_print-margin"
                      style="left: 4px; visibility: hidden;"
                    />
                  </div>
                  <div
                    class="ace_layer ace_marker-layer"
                  />
                  <div
                    class="ace_layer ace_text-layer"
                    style="padding: 0px 4px;"
                  />
                  <div
                    class="ace_layer ace_marker-layer"
                  />
                  <div
                    class="ace_layer ace_cursor-layer ace_hidden-cursors"
                  >
                    <div
                      class="ace_cursor"
                    />
                  </div>
                </div>
              </div>
              <div
                class="ace_scrollbar ace_scrollbar-v"
                style="display: none; width: 20px;"
              >
                <div
                  class="ace_scrollbar-inner"
                  style="width: 20px;"
                />
              </div>
              <div
                class="ace_scrollbar ace_scrollbar-h"
                style="display: none; height: 20px;"
              >
                <div
                  class="ace_scrollbar-inner"
                  style="height: 20px;"
                />
              </div>
              <div
                style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: hidden;"
              >
                <div
                  style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
                />
                <div
                  style="height: auto; width: auto; top: 0px; left: 0px; visibility: hidden; position: absolute; white-space: pre; overflow: visible;"
                >
                  XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="border "
      style="width: -562px;"
    >
      <div
        class="border  vertical"
        style="width: -562px; height: 384px; padding-bottom: 5px;"
      >
        <div
          class="item"
        >
          <div
            class="label"
          >
            unit test result
          </div>
          <div
            class="tape"
          >
            <div
              class="description"
            >
              1
            </div>
          </div>
        </div>
      </div>
      <div
        class="border "
        style="width: -562px; height: -384px;"
      >
        <div
          class="item"
        >
          <div
            class="label"
          >
            console
          </div>
          <div
            class="console-widget "
          >
            <div
              class="console"
            >
              <div
                class="text"
              >
                1
              </div>
            </div>
            <div
              class="console"
            >
              <div
                class="text"
              >
                2
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
